src/fast_array_utils/stats/_is_constant.py

```diff
@@ -0,0 +1,142 @@
+    axis
+        Axis to reduce over.
+
+    Returns
+    -------
+    If ``axis`` is :data:`None`, return if all values were constant.
+    Else returns a boolean array with :data:`True` representing constant columns/rows.
+
+    Example
+    -------
+    >>> a = np.array([[0, 1], [0, 0]])
+    >>> a
+    array([[0, 1],
+           [0, 0]])
+    >>> is_constant(a)
+    False
+    >>> is_constant(a, axis=0)
+    array([ True, False])
+    >>> is_constant(a, axis=1)
+    array([False,  True])
+
+    """
+    validate_axis(axis)
+    return _is_constant(a, axis=axis)
+
+
+@singledispatch
+def _is_constant(
+    a: NDArray[Any] | types.CSBase | types.DaskArray, /, *, axis: Literal[0, 1, None] = None
+) -> bool | NDArray[np.bool_]:  # pragma: no cover
+    raise NotImplementedError
+
+
+@_is_constant.register(np.ndarray)
+def _(a: NDArray[Any], /, *, axis: Literal[0, 1, None] = None) -> bool | NDArray[np.bool_]:
+    # Should eventually support nd, not now.
+    match axis:
+        case None:
+            return bool((a == a.flat[0]).all())
+        case 0:
+            return _is_constant_rows(a.T)
+        case 1:
+            return _is_constant_rows(a)
+
+
+def _is_constant_rows(a: NDArray[Any]) -> NDArray[np.bool_]:
+    b = np.broadcast_to(a[:, 0][:, np.newaxis], a.shape)
+    return (a == b).all(axis=1)  # type: ignore[no-any-return]
+
+
+@_is_constant.register(types.CSBase)
+def _(a: types.CSBase, /, *, axis: Literal[0, 1, None] = None) -> bool | NDArray[np.bool_]:
+    n_row, n_col = a.shape
+    if axis is None:
+        if len(a.data) == n_row * n_col:
+            return is_constant(a.data)  # type: ignore[no-any-return]
+        return bool((a.data == 0).all())
+    shape = (n_row, n_col) if axis == 1 else (n_col, n_row)
+    match axis, a.format:
+        case 0, "csr":
+            a = a.T.tocsr()
+        case 1, "csc":
+            a = a.T.tocsc()
+    return _is_constant_csr_rows(a.data, a.indptr, shape)
+
+
+@numba.njit(cache=True)
+def _is_constant_csr_rows(
+    data: NDArray[np.number[Any]],
+    indptr: NDArray[np.integer[Any]],
+    shape: tuple[int, int],
+) -> NDArray[np.bool_]:
+    n = len(indptr) - 1
+    result = np.ones(n, dtype=np.bool_)
+    for i in numba.prange(n):
+        start = indptr[i]
+        stop = indptr[i + 1]
+        val = data[start] if stop - start == shape[1] else 0
+        for j in range(start, stop):
+            if data[j] != val:
+                result[i] = False
+                break
+    return result
+
+
+@_is_constant.register(types.DaskArray)
+def _(a: types.DaskArray, /, *, axis: Literal[0, 1, None] = None) -> types.DaskArray:
+    if TYPE_CHECKING:
+        from dask.array.core import map_blocks
+    else:
+        from dask.array import map_blocks
+
+    if isinstance(a._meta, np.ndarray) and axis is None:  # noqa: SLF001
+        v = a[0, 0].compute()
+        return map_blocks(bool, (a == v).all(), meta=np.array([], dtype=bool))  # type: ignore[no-any-return,no-untyped-call]
```

Comment:
My guess is that you're mapping the `bool` value here so this will result in a bool.  I think dask expects the function to obey `meta`, certainly in the case of this sort of reduction operation

Reply:
You would expect `result = stats.is_constant(x, axis=None).compute() ` to be a numpy array?

Reply:
It’s never a numpy array. without the `map_blocks`, it’s a numpy scalar.

All the other overloads return a Python bool, so I added the `map_blocks` here for consistency.

Reply:
>  can you help me understand why this works?

Why would the dask tests not pass then? If you're telling it to return a `bool`, that's what it's doing.  I must be missing something

Reply:
https://github.com/scverse/fast-array-utils/blob/19a4187759fef400142d43a2e4196fcd3f750b10/tests/test_stats.py#L124

So we have this:

```py
x = np.array([
  [0, 0, 1, 1],
  [0, 0, 1, 1],
  [2, 2, 3, 3],
  [2, 2, 3, 3],
])
a = da.from_array(x, (2, 2))
assert a.blocks == [
  [np.repeat(0, (2, 2)), np.repeat(1, (2, 2))],
  [np.repeat(2, (2, 2)), np.repeat(3, (2, 2))],
]
```

I’d imagine that dask then runs `is_constant(block, axis=None)` on each block, resulting in

```py
is_constant(a, axis=None).blocks == [
  [True, True],
  [True, True],
]
```

… and then combines them somehow into `True`.

But clearly that doesn’t happen, so what happens instead and is that correct?

Reply:
I assume you're referring to the [second condition](https://github.com/scverse/fast-array-utils/pull/21/files#diff-43121ffc0d2cd17ea2e57432a5e4b87f6bab5937d96e1ff6d262fb6e63e7e824R138-R147) since the [first](https://github.com/scverse/fast-array-utils/pull/21/files#diff-43121ffc0d2cd17ea2e57432a5e4b87f6bab5937d96e1ff6d262fb6e63e7e824R133-R136) one makes sense (you call `(a == v).all()` which returns a dask array with `()` shape, and then map it to `bool`).

In this case, I think you're not hitting that condition with your test since you're only using dense.

Reply:
Checking it out locally, it still passes with sparse but does seem to materialize the whole array in memory.  So it might be best to do reduction here.  Why it's loading the whole thing into memory is not clear to me.

Reply:
Thanks! No idea either!